### PR TITLE
Handle synthesized top-level types in EnC symbol matcher

### PR DESCRIPTION
**Pull request by tmat**
The symbol matcher does not handle embedded attributes that the compiler emits to support features like `ref readonly`, nullable reference types, etc. The matcher needs to be aware of these attributes so that we can perform EnC on code that emits them.

This change comprises of several refactoring-only commits that clean up and simplify related code,
fix to https://github.com/dotnet/roslyn/issues/36926 and finally changes to symbol matcher to be able to match synthesized top-level types in potentially synthesized namespaces.

Also remove rude edits related to using `ref readonly` returns, `in` parameters and `ref` structs.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-compiler @jcouv @ivanbasov @agocke PTAL

**Comment by tmat** [comment]
@jinujoseph FYI

**Comment by jcouv** [comment]
Looking

File: src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs [hunk 1]
```diff
@@ -137,7 +136,46 @@ internal abstract class CommonPEModuleBuilder : Cci.IUnit, Cci.IModuleReference
...
+        /// </summary>
+        public abstract IEnumerable<Cci.INamespaceTypeDefinition> GetAdditionalTopLevelTypeDefinitions(EmitContext context);
+
+        /// <summary>
+        /// Anonymous types defined in the compilation.
+        /// </summary>
+        public abstract IEnumerable<Cci.INamespaceTypeDefinition> GetAnonymousTypeDefinitions(EmitContext context);
+
+        /// <summary>
+        /// Top-level embedded types (e.g. attribute types taht are not present in referenced assemblies).
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: >taht [](start = 59, length = 4)

Typo.

File: src/Compilers/Core/Portable/Emit/EditAndContinue/SymbolChanges.cs [hunk 2]
```diff
@@ -159,40 +166,41 @@ private SymbolChange GetChange(ISymbol symbol)
...
                     {
                         // If the definition did not exist in the previous generation, it was added.
-                        return SymbolChange.Added;
+                        return _definitionMap.DefinitionExists(definition) ? SymbolChange.None : SymbolChange.Added;
+                    }
+
+                    if (symbol is INamespace @namespace)
+                    {
+                        // If the namespace did not exist in the previous generation, it was added.
+                        // Otherwise the namespace may contains changes.
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/SymbolChanges.cs]: >contains [](start = 55, length = 8)

"contain"

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs [hunk 3]
```diff
@@ -50,27 +51,31 @@ internal sealed class CSharpSymbolMatcher : SymbolMatcher
...
             {
                 return (Cci.IDefinition)_symbols.Visit(symbol);
             }
 
             return _defs.VisitDef(definition);
         }
 
+        public override Cci.INamespace MapNamespace(Cci.INamespace @namespace)
+        {
+            Debug.Assert(@namespace is NamespaceSymbol);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: >Debug.Assert(@namespace is NamespaceSymbol); [](start = 12, length = 44)

Minor point: assert seems unnecessary.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs [hunk 4]
```diff
@@ -276,112 +281,174 @@ private void CreateEmbeddedAttributesIfNeeded(DiagnosticBag diagnostics)
...
-        private void CreateEmbeddedAttributeIfNeeded(
-            ref SynthesizedEmbeddedAttributeSymbol symbol,
+        private SynthesizedEmbeddedAttributeSymbol CreateParameterlessEmbeddedAttributeSymbol(string name, NamespaceSymbol containingNamespace, DiagnosticBag diagnostics)
+            => new SynthesizedEmbeddedAttributeSymbol(
+                    name,
+                    containingNamespace,
+                    SourceModule,
+                    baseType: GetWellKnownType(WellKnownType.System_Attribute, diagnostics));
+
+        private SynthesizedEmbeddedNullableAttributeSymbol CreateNullableAttributeSymbol(string name, NamespaceSymbol containingNamespace, DiagnosticBag diagnostics)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs]: >CreateNullableAttributeSymbol [](start = 59, length = 29)

Consider inlining these `CreateNullable*AttributeSymbol` methods as lambdas, so the delegates are cached.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs [hunk 5]
```diff
@@ -276,112 +281,174 @@ private void CreateEmbeddedAttributesIfNeeded(DiagnosticBag diagnostics)
...
-                    diagnostics);
+                    diagnostics,
+                    AttributeDescription.NullablePublicOnlyAttribute,
+                    CreateNullablePublicOnlyAttributeSymbol);
             }
         }
 
-        private void CreateEmbeddedAttributeIfNeeded(
-            ref SynthesizedEmbeddedAttributeSymbol symbol,
+        private SynthesizedEmbeddedAttributeSymbol CreateParameterlessEmbeddedAttributeSymbol(string name, NamespaceSymbol containingNamespace, DiagnosticBag diagnostics)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs]: >private SynthesizedEmbeddedAttributeSymbol [](start = 8, length = 42)

Consider declaring as:
```
private readonly Func<string, NamespaceSymbol, DiagnosticBag, SynthesizedEmbeddedAttributeSymbol> CreateParameterlessEmbeddedAttributeSymbol(...);
```

File: src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs [hunk 6]
```diff
@@ -9,6 +9,7 @@
 using System.Runtime.InteropServices;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Symbols;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs]: >Microsoft.CodeAnalysis.Symbols [](start = 6, length = 30)

Is this change needed?

File: src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+using System;
+using Microsoft.CodeAnalysis.Emit;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// Synthesized namespace that contains synthesized types or subnamespaces.
+    /// All its members are stored in a table on <see cref="CommonPEModuleBuilder"/>.
+    /// </summary>
+    internal sealed class SynthesizedNamespaceSymbol : NamespaceSymbol, IEquatable<SynthesizedNamespaceSymbol>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: >IEquatable<SynthesizedNamespaceSymbol> [](start = 72, length = 38)

Who uses this?

Comment (tmat) [reply src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: I followed the pattern of MissingNamespaceSymbol, which implements equality.

---
In reply to: [302225353](https://github.com/dotnet/roslyn/pull/36854#discussion_r302225353) [](ancestors = 302225353)

File: src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs (hunk 5 above)
Comment (tmat) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs]: This doesn't work since the lambda bodies reference instance members, so I'd need to move them to constructor. Not sure if it's worth the effort for saving a few delegate allocations per compilation emit.

I could store the first one in a local since it is used 4 times:



---
In reply to: [302215939](https://github.com/dotnet/roslyn/pull/36854#discussion_r302215939) [](ancestors = 302215939)

Comment (cston) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs]: Makes sense to store the first in a local and leave the methods as methods.

---
In reply to: [302235341](https://github.com/dotnet/roslyn/pull/36854#discussion_r302235341) [](ancestors = 302235341,302215939)

File: src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs (hunk 7 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: If there are no consumers of this interface, we should remove it.

---
In reply to: [302230781](https://github.com/dotnet/roslyn/pull/36854#discussion_r302230781) [](ancestors = 302230781,302225353)

File: src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            => ImmutableArray<NamedTypeSymbol>.Empty;
+
+        public override ImmutableArray<NamedTypeSymbol> GetTypeMembers(string name)
+            => ImmutableArray<NamedTypeSymbol>.Empty;
+
+        public override ImmutableArray<NamedTypeSymbol> GetTypeMembers(string name, int arity)
+            => ImmutableArray<NamedTypeSymbol>.Empty;
+
+        public override ImmutableArray<Symbol> GetMembers()
+            => ImmutableArray<Symbol>.Empty;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: >ImmutableArray<Symbol>.Empty [](start = 15, length = 28)

This seems incorrect in general.

If there are no callers of the `Get*Members(...)` methods, the methods should probably `throw ExceptionUtilities.Unreachable;` instead.

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs [hunk 9]
```diff
@@ -8,6 +8,7 @@
 using System.Threading;
 using Microsoft.CodeAnalysis.CSharp.Emit;
 using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Symbols;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs]: >Microsoft.CodeAnalysis.Symbols [](start = 6, length = 30)

Is this needed?

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicDefinitionMap.vb [hunk 10]
```diff
@@ -7,6 +7,7 @@ Imports Microsoft.CodeAnalysis.CodeGen
 Imports Microsoft.CodeAnalysis.Emit
 Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
 Imports Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+Imports Microsoft.CodeAnalysis.Symbols
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicDefinitionMap.vb]: >Microsoft.CodeAnalysis.Symbols [](start = 8, length = 30)

Please sort or remove if not needed.

File: src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs (hunk 8 above)
Comment (tmat) [reply src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: Why incorrect? The synthesized members are associated with the namespace via a look-aside table. The namespace does not have any declared members.

---
In reply to: [302277474](https://github.com/dotnet/roslyn/pull/36854#discussion_r302277474) [](ancestors = 302277474)

File: src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs (hunk 7 above)
Comment (tmat) [reply src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: Should I remove Equals and GetHashCode impl as well? If not then we should keep the interface, since that's a standard pattern how to implement equality.

---
In reply to: [302274715](https://github.com/dotnet/roslyn/pull/36854#discussion_r302274715) [](ancestors = 302274715,302230781,302225353)

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicDefinitionMap.vb (hunk 10 above)
Comment (tmat) [reply src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicDefinitionMap.vb]: Hmm, not sure why this was added everywhere...

---
In reply to: [302288682](https://github.com/dotnet/roslyn/pull/36854#discussion_r302288682) [](ancestors = 302288682)

File: src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs (hunk 7 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: We should remove the interface implementation only. No `Symbol` types implement `IEquatable<>`.

---
In reply to: [302289323](https://github.com/dotnet/roslyn/pull/36854#discussion_r302289323) [](ancestors = 302289323,302274715,302230781,302225353)

File: src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs (hunk 8 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: A caller will get a misleading result since these namespaces each have members. Better to fail fast.

---
In reply to: [302289122](https://github.com/dotnet/roslyn/pull/36854#discussion_r302289122) [](ancestors = 302289122,302277474)

Comment (tmat) [reply src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: I don't think it's misleading. Type symbols work like this as well. GetMembers on a type won't return synthesized nested types.
I *think* this might actually be called when enumerating top-level types to emit.

---
In reply to: [302290714](https://github.com/dotnet/roslyn/pull/36854#discussion_r302290714) [](ancestors = 302290714,302289122,302277474)

File: src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs (hunk 7 above)
Comment (tmat) [reply src/Compilers/CSharp/Portable/Symbols/SynthesizedNamespaceSymbol.cs]: OK. Sounds like a bug though.

---
In reply to: [302290074](https://github.com/dotnet/roslyn/pull/36854#discussion_r302290074) [](ancestors = 302290074,302289323,302274715,302230781,302225353)

**Comment by tmat** [comment]
@jcouv Thanks!
